src/main/java/com/google/devtools/build/lib/skyframe/BuildDriverFunction.java

```diff
@@ -339,10 +341,12 @@ public SkyValue compute(SkyKey skyKey, Environment env)
         NestedSet<Package> transitivePackagesForSymlinkPlanting =
             aspectValue.getTransitivePackages();
         if (transitivePackagesForSymlinkPlanting != null) {
-          postEventIfNecessary(
-              postedEventsTypes,
-              env,
-              TopLevelTargetReadyForSymlinkPlanting.create(transitivePackagesForSymlinkPlanting));
+          // Don't deduplicate this event per BuildDriverKey as the transitive packages can differ
```

Comment:
If the missing deduplication should turn out to be a concern, I could:
* Introduce enum constants up to some fixed number of aspects.
* Replace the inner `HashSet` with an `EnumSet` to potentially avoid some overhead elsewhere in this function. This may be a useful optimization in any case.

Reply:
I think deduplicating here is good idea, first and foremost to keep the behavior consistent: no duplicated events would be fired from `BuildDriverFunction`.

If we simply deduplicate by the `TopLevelTargetReadyForSymlinkPlanting` event, we could potentially end up with a race condition, though idk how likely:

- `aspect_A1.getTransitivePackages() == aspect_A2.getTransitivePackages() == pkgA`
- We send out `TopLevelTargetReadyForSymlinkPlanting(pkgA)`. `aspect_A1`'s execution is blocked on the symlinks being planted.
- `aspect_A2` also wants to send out `TopLevelTargetReadyForSymlinkPlanting(pkgA)`, and gets a cache hit.
- `aspect_A2` carries on with execution => missing files.


So perhaps you'll have to deduplicate it per-Aspect.

Reply:
> If we simply deduplicate by the `TopLevelTargetReadyForSymlinkPlanting` event, we could potentially end up with a race condition...

In the scenario you describe above, wouldn't `aspect_A2` be blocked on the `lazilyPlantSymlink`s event handler, which explicitly isn't annotated with `@AllowConcurrentEvents`, until `aspect_A1`'s event handling has finished, which includes creating the symlinks? Only after that should `aspect_A2` even enter the logic that is short circuited with a cache hit. If this wasn't the case, wouldn't the current logic already be incorrect even without considering multiple aspects attached to a single target?

I will look into how to deduplicate events efficiently, even if only for consistency. Is there a bound on the number of aspects that can be applied to a single target?